crates/bevy_ecs/src/schedule_v3/schedule.rs

```diff
@@ -0,0 +1,1001 @@
+        }
+
+        if !self.dependency.graph.contains_node(id) {
+            self.dependency.graph.add_node(id);
+        }
+
+        for (edge_kind, other) in dependencies
+            .into_iter()
+            .map(|(edge_kind, set)| (edge_kind, self.system_set_ids[&set]))
+        {
+            let (lhs, rhs) = match edge_kind {
+                DependencyEdgeKind::Before => (id, other),
+                DependencyEdgeKind::After => (other, id),
+            };
+            self.dependency.graph.add_edge(lhs, rhs, ());
+        }
+
+        match ambiguous_with {
+            Ambiguity::Check => (),
+            Ambiguity::IgnoreWithSet(ambigous_with) => {
+                for other in ambigous_with
+                    .into_iter()
+                    .map(|set| self.system_set_ids[&set])
+                {
+                    self.ambiguous_with.add_edge(id, other, ());
+                }
+            }
+            Ambiguity::IgnoreAll => {
+                self.ambiguous_with_all.insert(id);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn initialize(&mut self, world: &mut World) {
+        for (id, uninit) in self.uninit.drain(..) {
+            match uninit {
+                UninitNode::System(mut system, mut conditions) => {
+                    debug_assert!(id.is_system());
+                    system.initialize(world);
+                    self.systems.insert(id, system);
+                    for condition in conditions.iter_mut() {
+                        condition.initialize(world);
+                    }
+                    self.conditions.insert(id, conditions);
+                }
+                UninitNode::SystemSet(mut conditions) => {
+                    debug_assert!(id.is_set());
+                    for condition in conditions.iter_mut() {
+                        condition.initialize(world);
+                    }
+                    self.conditions
+                        .entry(id)
+                        .or_insert_with(Vec::new)
+                        .append(&mut conditions);
+                }
+            }
+        }
+    }
+
+    fn build_dependency_graph(&mut self) -> Result<(), BuildError> {
+        // check hierarchy for cycles
+        let hier_scc = tarjan_scc(&self.hierarchy.graph);
+        if self.contains_cycles(&hier_scc) {
+            self.report_cycles(&hier_scc);
+            return Err(BuildError::HierarchyCycle);
+        }
+
+        self.hierarchy.topsort = hier_scc.into_iter().flatten().rev().collect::<Vec<_>>();
+
+        let hier_results = check_graph(&self.hierarchy.graph, &self.hierarchy.topsort);
+        if self.contains_hierarchy_conflicts(&hier_results.transitive_edges) {
+            self.report_hierarchy_conflicts(&hier_results.transitive_edges);
+            return Err(BuildError::HierarchyConflict);
+        }
+
+        // check dependencies for cycles
+        let dep_scc = tarjan_scc(&self.dependency.graph);
```

Comment:
Can we clone instead of repeating the computation?

Reply:
What computation was repeated?

Reply:
tarjan_scc here and above.

Reply:
It's being used on different graphs.

Reply:
Oh yep, totally missed that. This can be resolved.